File: five_problems.rb
# Problem 1
# Write three functions that compute the sum of the numbers in a given list using a for-loop, a while-loop, and recursion.
def sum1(num_array)
  sum = 0
  for num in num_array do
    sum+=num
  end
  sum
end

def sum2(num_array)
  sum = 0
  i = 0
  while i<num_array.length
    sum += num_array[i]
    i+=1
  end
end

def sum3(num_array)
  len = num_array.length
  if len == 1
    return num_array[0]
  else
    sum3(num_array[0..len/2]) + sum3(num_array[len/2..len-1])
  end
end

# Problem 2
#
# Write a function that combines two lists by alternatingly taking elements. For example: given the two lists [a, b, c] and [1, 2, 3], the function should return [a, 1, b, 2, c, 3].

def combonie(a1, a2)
  i=0
  combo = []
  while !(a1[i].nil? and a2[i].nil?)
    unless a1[i].nil?
      combo << a1[i]
    end
    unless a2[i].nil?
      combo << a2[i]
    end
    i+=1
  end
  combo
end

# Problem 3
# Write a function that computes the list of the first 100 Fibonacci numbers. By definition, the first two numbers in the Fibonacci sequence are 0 and 1, and each subsequent number is the sum of the previous two. As an example, here are the first 10 Fibonnaci numbers: 0, 1, 1, 2, 3, 5, 8, 13, 21, and 34.

# recursion version (slower than non recursive)
def fib(n)
  answer = []
  return n if n<= 1
  fib(n-1) + fib(n-2)
end

def fib(digit)
  if digit == 1
    return [0]
  end
  answer = [0,1]
  i=0
  (digit-2).times do
    num = answer[i] + answer[i+1]
    answer << num
    i+=1
  end
  answer
end

File: five_problems.js
// Write three functions that compute the sum of the numbers in a given list using a for-loop, a while-loop, and recursion.

function for_loop_summer(num_list) {
  var len = num_list.length;
  var sum = 0;
  for (i=0; i<len; i++) {
    sum += num_list[i];
  }
  return sum;
}

function while_loop_summer(num_list) {
  var i = 0;
  var sum = 0;
  while (num_list[i] !== undefined) {
    sum += num_list[i];
    i++;
  }
  return sum;
}

function recursion_summer(num_list) {
  var len = num_list.length;
  if (len == 1) {
    return num_list[0];
  }
  return recursion_summer(num_list.slice(0,1)) + recursion_summer(num_list.slice(1,len));
}

// Write a function that combines two lists by alternatingly taking elements. For example: given the two lists [a, b, c] and [1, 2, 3], the function should return [a, 1, b, 2, c, 3].

function list_weaver(list1, list2) {
  if (list1.length > list2.length) {
    return combine_lists(list1, list2);
  } else {
    return combine_lists(list2,list1);
  }
}
function combine_lists(long, short) {
  var i = 0;
  var new_list = [];
  while(short[i] != undefined) {
    new_list.push(short[i])
    new_list.push(long[i])
    i ++
  }
  return new_list.concat(long.slice(i,long.length));
}
